db/sql/sql.go

```diff
@@ -932,6 +998,72 @@ func (tx *Transaction) ListContext(ctx context.Context, s *schema.Schema, filter
 	return tx.executeSelect(ctx, sc, sql, args)
 }
 
+func (tx *CachedTransaction) List(s *schema.Schema, filter transaction.Filter, options *transaction.ViewOptions, pg *pagination.Paginator) (list []*schema.Resource, total uint64, err error) {
+	return tx.ListContext(context.Background(), s, filter, options, pg)
+}
+
+//List resources in the db
+func (tx *CachedTransaction) ListContext(ctx context.Context, s *schema.Schema, filter transaction.Filter, options *transaction.ViewOptions, pg *pagination.Paginator) (list []*schema.Resource, total uint64, err error) {
+	sc := &selectContext{
+		schema:    s,
+		filter:    filter,
+		join:      true,
+		paginator: pg,
+	}
+	if options != nil {
+		sc.fields = normFields(options.Fields, s)
+		sc.join = options.Details
+	}
+
+	sql, args, err := buildSelect(sc)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	list, total, err, errCache := tx.getCached(s.ID, sc, sql, args)
+
+	if errCache != nil {
+		if tx.TxInterface == nil {
+		}
+		list, total, err = tx.TxInterface.ListContext(ctx, s, filter, options, pg)
+		tx.saveCache(s.ID, sc, sql, args, list, total, err)
+	}
+	return
+}
+
+func hash(s string) uint32 {
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	return h.Sum32()
+}
+
+func (tx *CachedTransaction) createKey(schemeID string, sc *selectContext, sql string, args []interface{}) uint32 {
+	sqlFormat := strings.Replace(sql, "?", "%s", -1)
+	query := fmt.Sprintf(sqlFormat, args...)
+	res := fmt.Sprintf("%s %s %v %v %v",schemeID, query, sc.join, sc.filter, sc.fields);
+	return hash(res)
+}
+
+func (tx *CachedTransaction) getCached(schemeID string, sc *selectContext, sql string, args []interface{}) (list []*schema.Resource, total uint64, errList error, errCached error) {
+	key := tx.createKey(schemeID, sc, sql, args)
+	res, present := tx.QueryCache[key]
+	if present {
```

Comment:
I'd rather go the defensive way and check for hash collisions

Reply:
Ok, lets assume that we have a collision, what should we do? Only decent solution I can think of is to add the original query to the cache and check it always, after extracting anything from cache. And what then? we can just assume that we only cache one query with the given key or add the key at the end of the query to create another key and store the new request under it.

Reply:
in fact we could just use the plain-text query as a key instead of the hash in the `QueryCache` (which is a hashmap itself, so apart from small memory optimization there's not much added benefit)

Reply:
memory + speed as map operations on long strings can be slow... but we won't have too many records there so it should be fine anyway